File: AlgorithmicToolbox/week4_divide_and_conquer/3_majority_element/3_majority_element.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using std::vector;

/*
Majority rule is a decision rule that selects the alternative which has a majority,
that is, more than half the votes.
Given a sequence of elements a1, a2, a3, . . ., an, you would like to check whether
it contains an element that appears more than n/2 times.
*/

int get_majority_element(vector<int>& a, int left, int right) {
    if (left == right) return a[left];
    
    int mid =(right + left) / 2;

    int left_majority = get_majority_element(a, left, mid);
    int right_majority = get_majority_element(a, mid + 1, right);

    if (left_majority == right_majority) return left_majority;

    int left_count = 0, right_count = 0;
    for (int i = 0; i < a.size(); ++i) {
        if (a[i] == left_majority) left_count += 1;
        if (a[i] == right_majority) right_count += 1;
    }

    return (left_count > right_count) ? left_majority : right_majority;
}

int majority_element(vector<int>& a) {
    int majority_elem = get_majority_element(a, 0, a.size()-1);
    int count = 0;

    for (auto elem : a) {
        if (elem == majority_elem) count += 1;
    }

    return (count > a.size() / 2) ? 1 : 0;
}


int main() {
    int n;
    std::cin >> n;
    vector<int> a(n);
    for (size_t i = 0; i < a.size(); ++i) {
        std::cin >> a[i];
    }
    std::cout << majority_element(a) << '\n';

    return 0;
}

File: AlgorithmicToolbox/week4_divide_and_conquer/3_majority_element/3_majority_element.py
#!/usr/bin/env python
# coding: utf-8

# In[1]:


# Majority element
"""
Majority rule is a decision rule that selects the alternative which has a majority,
that is, more than half the votes.
Given a sequence of elements a1, a2, a3, . . ., an, you would like to check whether
it contains an element that appears more than n/2 times. 
"""

def majority_element_naive(elements):
    assert len(elements) <= 10 ** 5
    for e in elements:
        if elements.count(e) > len(elements) / 2:
            return 1

    return 0


def majority_element(elements):
    assert len(elements) <= 10 ** 5
    
    majority_elem = get_majority(elements, 0, len(elements)-1)
    
    count = 0
    for elem in elements:
        if elem == majority_elem:
            count += 1
    return 1 if count > (len(elements)//2) else 0


def get_majority(elements, low, high):
    
    if low == high:
        return elements[low]
    
    mid = low + (high-low)//2
    
    left_majority = get_majority(elements, low, mid)
    right_majority = get_majority(elements, mid+1, high)
    
    if left_majority == right_majority:
        return left_majority
    
    # otherwise, count each element and return the "winner".
    left_count = sum(1 for i in range(low, high+1) if elements[i] == left_majority)
    right_count = sum(1 for i in range(low, high+1) if elements[i] == right_majority)
    
    return left_majority if left_count > right_count else right_majority
    


if __name__ == '__main__':
    input_n = int(input())
    input_elements = list(map(int, input().split()))
    assert len(input_elements) == input_n
    print(majority_element(input_elements))


# In[ ]:




